### PR TITLE
add Leaflet.heatmap

### PR DESCRIPTION
**Pull request by timelyportfolio**
In response to #137, add [Leaflet.heat]() to `leaflet`.  While I think the implementation is complete, I would like some input on how I can make the `gradient` argument more `R`-like or consistent with `leaflet`.  As of now, it looks like this

```
 list(
      '0.2' = '#ffffb2',
      '0.4' = '#fd8d3c',
      '0.6' = '#fd8d3c',
      '0.8' = '#f03b20',
      '1' = '#bd0026'
  )
```

Some examples are in the `./inst/examples/heatmap.r` directory and also included in the documentation.  I'll paste here also.

```r
library(leaflet)

# match the heatmap demo from Leaflet.heat
# https://github.com/Leaflet/Leaflet.heat/blob/gh-pages/demo/index.html

# get the data to match exactly
addressPoints <- readLines(
  "http://leaflet.github.io/Leaflet.markercluster/example/realworld.10000.js"
)
addressPoints <- apply(
  jsonlite::fromJSON(
    sprintf("[%s]",
      paste0(
        addressPoints[4:(length(addressPoints)-1)]
        ,collapse=""
      )
    )
  )
  ,MARGIN = 2
  ,as.numeric
)

# create our heatmap
leaf <- leaflet() %>%
  setView( 175.475,-37.87, 12 ) %>%
  addHeatmap(addressPoints)

leaf

# customize our heatmap with options
library(RColorBrewer)
pal <- brewer.pal(9,"BuPu")

leaf <- leaflet() %>%
  setView( 175.475,-37.87, 12 ) %>%
  addHeatmap(
    addressPoints
    ,blur = 50
    # using example
    # https://esri.github.io/esri-leaflet/examples/styling-heatmaps.html
    ,gradient = list(
      '0.2' = '#ffffb2',
      '0.4' = '#fd8d3c',
      '0.6' = '#fd8d3c',
      '0.8' = '#f03b20',
      '1' = '#bd0026'
    )
  )

leaf

# replicate the example provided by
#   http://www.d3noob.org/2014/02/generate-heatmap-with-leafletheat-and.html

earthquakes <- readLines(
  "http://bl.ocks.org/d3noob/raw/8973028/2013-earthquake.js"
)
earthquakes <- apply(
  jsonlite::fromJSON(
    sprintf("[%s]",
            paste0(
              earthquakes[5:(length(earthquakes)-1)]
              ,collapse=""
            )
    )
  )
  ,MARGIN = 2
  ,as.numeric
)

leaflet() %>%
  addTiles() %>%
  setView( 174.146, -41.5546, 10 ) %>%
  addHeatmap(
    earthquakes,
    radius = 20,
    blur = 15,
    maxZoom = 17
  )

#  using data(quakes)
data(quakes)
quakes_mat <- matrix(t(quakes[,c(1:2,4)]),ncol=3,byrow=T)
leaflet() %>%
  addTiles( ) %>%
  setView( 178, -20, 5 ) %>%
  addHeatmap( quakes_mat, blur = 30, max = 0.05, radius = 30 )

# to remove the heatmap
leaf %>% clearHeatmap()
```

image

### CODE REVIEW

**Comment by jcheng5** [comment]
Is the range of the gradient always 0-1?

**Comment by jcheng5** [comment]
If the range of the gradient is always 0-1, then instead of a gradient you could take a color function, like other leaflet functions do.

```r
leaflet() %>%
  addTiles( ) %>%
  setView( 178, -20, 5 ) %>%
  addHeatmap( quakes_mat, blur = 30, max = 0.05, radius = 30,
    color = colorNumeric("RdYlBu", 0:1)
  )
```

Then inside `addHeatmap`, do something like

```r
gradient <- as.list(color(0:20 / 20))
names(gradient) <- as.character(0:20 / 20)
```

File: inst/examples/heatmap.R [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  addHeatmap(
+    earthquakes,
+    radius = 20,
+    blur = 15,
+    maxZoom = 17
+  )
+
+#  using data(quakes)
+data(quakes)
+quakes_mat <- matrix(t(quakes[,c(1:2,4)]),ncol=3,byrow=T)
```
Comment (yihui) [inline inst/examples/heatmap.R]: Could you change `T` to `TRUE` here?

Comment (timelyportfolio) [inline inst/examples/heatmap.R]: embarrassed that I let that through;  changed

**Comment by jcheng5** [comment]
Kenton, can I trouble you to not use the leading-comma style in this repo?

```r
foo(
  bar
  ,baz
  ,qux
)
```

I generally try to be extremely laissez-faire about code formatting style, but this is going too far afield IMHO. :)

**Comment by jcheng5** [comment]
Other than those comments, this looks great! Thanks!

**Comment by jcheng5** [comment]
Oh sorry, I missed something quite obvious. If latlngs are just latitude/longitude pairs, then we don't pass them like this in leaflet. See the usage of derivePoints in e.g. addMarkers, which takes separate lng/lat parameters (and unfortunately it is lng then lat).

**Comment by jcheng5** [comment]
Just to be crystal clear: addHeatmap should have lng=NULL, lat=NULL, and data=getMapData(map) parameters. Then call

```r
    pts = derivePoints(data, lng, lat, missing(lng), missing(lat), 
        "addHeatmap")
```

and pass `pts$lat` and `pts$lng` to invokeMethod.

**Comment by timelyportfolio** [comment]
@jcheng5, hope it is closer now.  I tried to follow what I found on the `JavaScript` and `R` side for consistency.

**Comment by timelyportfolio** [comment]
Not sure that is how the `max` works.  By leaving `0:1`, `max` is the only way to scale the intensity, so going to leave `0:1`.  The `alt` trick didn't work for me.  Leaflet.heat seems to want an array unless I'm just screwing up.  That is great to know on the auto-resolve.  Neat little bit of sugar.

**Comment by jcheng5** [comment]
@timelyportfolio Do you have time to vchat some time today or Monday? I'm very unclear on the semantics of intensity vs. max, even after studying the source code of Leaflet.Heat and simpleheat. Frankly I'm not sure the Leaflet.Heat behavior is correct/desirable. Can I sketch out what I'm seeing, you can tell me your understanding, and we'll see if we can make sense of it?

**Comment by timelyportfolio** [comment]
@jcheng5 happy to vchat, but I have to admit, I am at a loss also.  It is odd and not documented.  All I can determine is that reducing the `max` increases the color intensity.  Free anytime.  What is your preferred method?

**Comment by timelyportfolio** [comment]
@jcheng5 this `v` transform in [line](https://github.com/Leaflet/Leaflet.heat/blob/gh-pages/src/HeatLayer.js#L131) means our max will be much less than `1`.  Locally, I've added the non-minified.  If it helps, I can commit to a diff branch for debugging.

**Comment by timelyportfolio** [comment]
Take for example the `quakes` example and supply a multiplier to the `mag`.

```
data(quakes)
data(quakes)

leaflet(quakes) %>%
  addTiles( ) %>%
  setView( 178, -20, 5 ) %>%
  addHeatmap( lng = ~long, intensity = ~mag,
              blur = 5, radius = 2,
              gradient = "Greys" )
```

yields a `max` in the `data` supplied to simpleheat of 0.051.

```
leaflet(quakes) %>%
  addTiles( ) %>%
  setView( 178, -20, 5 ) %>%
  addHeatmap( lng = ~long, intensity = ~mag*1000,
              blur = 5, radius = 2,
              gradient = "Greys" )
```

gives us a `max` of 1.

I guess if we want to be smart :) we could calculate the `k` on the `R` side and try to scale the intensity values to [0,1].

**Comment by timelyportfolio** [comment]
With your pull request to Leaflet.heat,I just discovered this very interesting and useful side effect.  We can easily create dot maps.  As an example,

```
library(leaflet)

# match the heatmap demo from Leaflet.heat
# https://github.com/Leaflet/Leaflet.heat/blob/gh-pages/demo/index.html

# get the data to match exactly
addressPoints <- readLines(
  "http://leaflet.github.io/Leaflet.markercluster/example/realworld.10000.js"
)
addressPoints <- apply(
  jsonlite::fromJSON(
    sprintf("[%s]",
      paste0(
        addressPoints[4:(length(addressPoints)-1)]
        ,collapse=""
      )
    )
  )
  ,MARGIN = 2
  ,as.numeric
)

addressPoints <- data.frame( addressPoints )
colnames( addressPoints ) <- c( "lat", "lng", "value" )

leaflet( addressPoints ) %>% addTiles() %>%
  setView( 175.475,-37.87, 12 ) %>%
  addHeatmap(intensity=~value, radius = 2, blur = 0, gradient = blues9[8:9] )
```

image

**Comment by timelyportfolio** [comment]
@jcheng5, there seems to be no movement on your pull request https://github.com/Leaflet/Leaflet.heat/pull/32.  I'm very much in favor of using your pull for this plugin for the purposes of this `leaflet` htmlwidget and incorporating future use cases, examples, and tests.  Otherwise, this seems like it might languish for a while, and I think this is too valuable to sit.

Thanks so much for all your help with this and finding what I think is a very good solution.

**Comment by smartinsightsfromdata** [comment]
@timelyportfolio @jcheng5 is this going ahead?  

It is a very useful plugin...

**Comment by bhaskarvk** [comment]
+1 will be nice to have.

**Comment by kezhan** [comment]
This really is an awesome feature ! I noticed that it is not yet integrated into master. Are you planning to do so ? Thank you all.

**Comment by odvratnozgodan** [comment]
Hi, great feature, been looking for some time for an interactive map with heatmap support. Do you know when will it be merged into the master branch?

**Comment by bhaskarvk** [comment]
I'm closing this. I'm moving all new plugins to leaflet.extras package.

**Comment by bhaskarvk** [comment]
Heatmap Support via an alternative but more performant plugin https://github.com/ursudio/webgl-heatmap-leaflet Now in trunk of https://github.com/bhaskarvk/leaflet.extras/

**Pull request closed without merging**